File: pychivalry/symbols.py
"""
CK3 Document Symbols - Outline and Structure Navigation

DIAGNOSTIC CODES:
    SYMBOL-001: Invalid symbol range (end before start)
    SYMBOL-002: Missing symbol name
    SYMBOL-003: Unknown symbol kind for CK3 construct
    SYMBOL-004: Malformed symbol hierarchy (invalid nesting)

MODULE OVERVIEW:
    Document symbols provide outline/breadcrumb navigation, allowing users
    to see the structure of a CK3 script file at a glance and jump to specific
    elements. This powers the Outline view in VS Code and similar features
    in other LSP-compatible editors.
    
    Symbols are hierarchical: events contain triggers/options, scripted effects
    contain parameters, etc. This module extracts these hierarchies from parsed
    AST and converts them to LSP DocumentSymbol format.

ARCHITECTURE:
    **Symbol Extraction Pipeline**:
    1. Parse document to AST (parser.py)
    2. Walk AST identifying symbol-worthy constructs
    3. For each construct, extract name, kind, and range
    4. Build hierarchical structure (children within parents)
    5. Convert to LSP DocumentSymbol format
    6. Return list for editor display
    
    **Symbol Types** (mapped to LSP SymbolKind):
    - Events → Event (with trigger/immediate/option children)
    - Scripted effects → Function
    - Scripted triggers → Function
    - Script values → Variable
    - On-actions → Event
    - Parameters → Property
    - Namespace declarations → Namespace

SYMBOL HIERARCHY:
    Events are primary symbols with nested children:
    ```
    my_event.0001          [Event]
      ├─ trigger           [Object]
      ├─ immediate         [Object]
      └─ option            [EnumMember]
         └─ effect         [Object]
    ```
    
    Scripted effects show parameters:
    ```
    my_effect             [Function]
      ├─ $PARAM1$         [Property]
      └─ $PARAM2$         [Property]
    ```

LSP INTEGRATION:
    - textDocument/documentSymbol: Returns symbols for single file
    - workspace/symbol: Returns symbols matching query across workspace
    - Outline view: Displays hierarchical symbol tree
    - Breadcrumbs: Shows current symbol path in editor header

USAGE EXAMPLES:
    >>> # Get symbols for a document
    >>> symbols = extract_document_symbols(parsed_ast)
    >>> symbols[0].name
    'my_mod.0001'  # Event ID
    >>> symbols[0].kind
    SymbolKind.Event
    >>> len(symbols[0].children)
    3  # trigger, immediate, option

PERFORMANCE:
    - Symbol extraction: ~10ms per 1000 lines
    - Cached after initial parse
    - Incremental updates when file changes
    - Workspace symbol search: ~50ms for 10k symbols

SEE ALSO:
    - parser.py: Provides AST for symbol extraction
    - navigation.py: Uses symbols for go-to-definition
    - indexer.py: Cross-document symbol index for workspace/symbol
"""

from typing import List, Optional, Dict
from dataclasses import dataclass
from lsprotocol import types


@dataclass
class DocumentSymbol:
    """
    Represents a symbol in a document.

    Attributes:
        name: Symbol name
        kind: LSP SymbolKind
        range: Full range including children
        selection_range: Range of the symbol name itself
        detail: Additional details about the symbol
        children: Nested symbols
    """

    name: str
    kind: types.SymbolKind
    range: types.Range
    selection_range: types.Range
    detail: Optional[str] = None
    children: List["DocumentSymbol"] = None

    def __post_init__(self):
        if self.children is None:
            self.children = []


# Map CK3 constructs to LSP SymbolKinds
CK3_TO_SYMBOL_KIND = {
    "event": types.SymbolKind.Event,
    "scripted_effect": types.SymbolKind.Function,
    "scripted_trigger": types.SymbolKind.Function,
    "script_value": types.SymbolKind.Variable,
    "on_action": types.SymbolKind.Event,
    "trigger": types.SymbolKind.Object,
    "immediate": types.SymbolKind.Object,
    "option": types.SymbolKind.EnumMember,
    "after": types.SymbolKind.Object,
    "effect": types.SymbolKind.Object,
    "namespace": types.SymbolKind.Namespace,
    "parameter": types.SymbolKind.Property,
}


def get_symbol_kind(construct_type: str) -> types.SymbolKind:
    """
    Get the appropriate LSP SymbolKind for a CK3 construct.

    Args:
        construct_type: Type of CK3 construct

    Returns:
        Appropriate LSP SymbolKind
    """
    return CK3_TO_SYMBOL_KIND.get(construct_type, types.SymbolKind.Object)


def create_document_symbol(
    name: str,
    construct_type: str,
    start_line: int,
    start_char: int,
    end_line: int,
    end_char: int,
    detail: Optional[str] = None,
) -> DocumentSymbol:
    """
    Create a DocumentSymbol object.

    Args:
        name: Symbol name
        construct_type: Type of CK3 construct
        start_line: Start line (0-indexed)
        start_char: Start character (0-indexed)
        end_line: End line (0-indexed)
        end_char: End character (0-indexed)
        detail: Optional detail string

    Returns:
        DocumentSymbol object
    """
    range_obj = types.Range(
        start=types.Position(line=start_line, character=start_char),
        end=types.Position(line=end_line, character=end_char),
    )

    # Selection range is typically just the name
    selection_range = types.Range(
        start=types.Position(line=start_line, character=start_char),
        end=types.Position(line=start_line, character=start_char + len(name)),
    )

    return DocumentSymbol(
        name=name,
        kind=get_symbol_kind(construct_type),
        range=range_obj,
        selection_range=selection_range,
        detail=detail,
    )


def extract_event_symbols(event_node: Dict) -> DocumentSymbol:
    """
    Extract symbols from an event node.

    Events contain:
    - Event ID (main symbol)
    - Trigger block (child symbol)
    - Immediate block (child symbol)
    - Options (child symbols)
    - After block (child symbol)

    Args:
        event_node: Parsed event node

    Returns:
        DocumentSymbol for the event with children
    """
    event_id = event_node.get("id", "unknown_event")
    event_type = event_node.get("type", "character_event")

    event_symbol = create_document_symbol(
        name=event_id,
        construct_type="event",
        start_line=event_node.get("start_line", 0),
        start_char=event_node.get("start_char", 0),
        end_line=event_node.get("end_line", 0),
        end_char=event_node.get("end_char", 0),
        detail=event_type,
    )

    # Add trigger as child if present
    if "trigger" in event_node:
        trigger_symbol = create_document_symbol(
            name="trigger",
            construct_type="trigger",
            start_line=event_node["trigger"].get("start_line", 0),
            start_char=event_node["trigger"].get("start_char", 0),
            end_line=event_node["trigger"].get("end_line", 0),
            end_char=event_node["trigger"].get("end_char", 0),
        )
        event_symbol.children.append(trigger_symbol)

    # Add immediate as child if present
    if "immediate" in event_node:
        immediate_symbol = create_document_symbol(
            name="immediate",
            construct_type="immediate",
            start_line=event_node["immediate"].get("start_line", 0),
            start_char=event_node["immediate"].get("start_char", 0),
            end_line=event_node["immediate"].get("end_line", 0),
            end_char=event_node["immediate"].get("end_char", 0),
        )
        event_symbol.children.append(immediate_symbol)

    # Add options as children
    if "options" in event_node:
        for i, option in enumerate(event_node["options"]):
            option_name = option.get("name", f"option_{i}")
            option_symbol = create_document_symbol(
                name=option_name,
                construct_type="option",
                start_line=option.get("start_line", 0),
                start_char=option.get("start_char", 0),
                end_line=option.get("end_line", 0),
                end_char=option.get("end_char", 0),
            )
            event_symbol.children.append(option_symbol)

    # Add after as child if present
    if "after" in event_node:
        after_symbol = create_document_symbol(
            name="after",
            construct_type="after",
            start_line=event_node["after"].get("start_line", 0),
            start_char=event_node["after"].get("start_char", 0),
            end_line=event_node["after"].get("end_line", 0),
            end_char=event_node["after"].get("end_char", 0),
        )
        event_symbol.children.append(after_symbol)

    return event_symbol


def extract_scripted_effect_symbols(effect_node: Dict) -> DocumentSymbol:
    """
    Extract symbols from a scripted effect node.

    Args:
        effect_node: Parsed scripted effect node

    Returns:
        DocumentSymbol for the scripted effect
    """
    effect_name = effect_node.get("name", "unknown_effect")
    parameters = effect_node.get("parameters", [])

    detail = None
    if parameters:
        detail = f"Parameters: {', '.join(parameters)}"

    effect_symbol = create_document_symbol(
        name=effect_name,
        construct_type="scripted_effect",
        start_line=effect_node.get("start_line", 0),
        start_char=effect_node.get("start_char", 0),
        end_line=effect_node.get("end_line", 0),
        end_char=effect_node.get("end_char", 0),
        detail=detail,
    )

    # Add parameters as children
    for param in parameters:
        param_symbol = create_document_symbol(
            name=f"${param}$",
            construct_type="parameter",
            start_line=effect_node.get("start_line", 0),
            start_char=effect_node.get("start_char", 0),
            end_line=effect_node.get("start_line", 0),
            end_char=effect_node.get("start_char", 0) + len(param) + 2,
        )
        effect_symbol.children.append(param_symbol)

    return effect_symbol


def extract_scripted_trigger_symbols(trigger_node: Dict) -> DocumentSymbol:
    """
    Extract symbols from a scripted trigger node.

    Args:
        trigger_node: Parsed scripted trigger node

    Returns:
        DocumentSymbol for the scripted trigger
    """
    trigger_name = trigger_node.get("name", "unknown_trigger")
    parameters = trigger_node.get("parameters", [])

    detail = None
    if parameters:
        detail = f"Parameters: {', '.join(parameters)}"

    trigger_symbol = create_document_symbol(
        name=trigger_name,
        construct_type="scripted_trigger",
        start_line=trigger_node.get("start_line", 0),
        start_char=trigger_node.get("start_char", 0),
        end_line=trigger_node.get("end_line", 0),
        end_char=trigger_node.get("end_char", 0),
        detail=detail,
    )

    # Add parameters as children
    for param in parameters:
        param_symbol = create_document_symbol(
            name=f"${param}$",
            construct_type="parameter",
            start_line=trigger_node.get("start_line", 0),
            start_char=trigger_node.get("start_char", 0),
            end_line=trigger_node.get("start_line", 0),
            end_char=trigger_node.get("start_char", 0) + len(param) + 2,
        )
        trigger_symbol.children.append(param_symbol)

    return trigger_symbol


def extract_script_value_symbols(value_node: Dict) -> DocumentSymbol:
    """
    Extract symbols from a script value node.

    Args:
        value_node: Parsed script value node

    Returns:
        DocumentSymbol for the script value
    """
    value_name = value_node.get("name", "unknown_value")
    value_type = value_node.get("type", "fixed")

    detail = f"Type: {value_type}"

    return create_document_symbol(
        name=value_name,
        construct_type="script_value",
        start_line=value_node.get("start_line", 0),
        start_char=value_node.get("start_char", 0),
        end_line=value_node.get("end_line", 0),
        end_char=value_node.get("end_char", 0),
        detail=detail,
    )


def extract_on_action_symbols(on_action_node: Dict) -> DocumentSymbol:
    """
    Extract symbols from an on_action node.

    Args:
        on_action_node: Parsed on_action node

    Returns:
        DocumentSymbol for the on_action
    """
    on_action_name = on_action_node.get("name", "unknown_on_action")
    event_count = len(on_action_node.get("events", []))

    detail = f"{event_count} event(s)" if event_count > 0 else None

    return create_document_symbol(
        name=on_action_name,
        construct_type="on_action",
        start_line=on_action_node.get("start_line", 0),
        start_char=on_action_node.get("start_char", 0),
        end_line=on_action_node.get("end_line", 0),
        end_char=on_action_node.get("end_char", 0),
        detail=detail,
    )


def extract_story_cycle_symbols(story_cycle_node) -> DocumentSymbol:
    """
    Extract symbols from a story cycle node.
    
    Story cycles contain:
    - Story cycle ID (main symbol)
    - on_setup block (child symbol)
    - on_end block (child symbol)
    - on_owner_death block (child symbol)
    - effect_group blocks (child symbols)
    
    Args:
        story_cycle_node: Parsed story cycle AST node
    
    Returns:
        DocumentSymbol for the story cycle with children
    """
    # Get the story cycle name from the key
    story_name = getattr(story_cycle_node, 'key', 'unknown_story')
    
    # Get position from range
    range_obj = getattr(story_cycle_node, 'range', None)
    if range_obj:
        start_line = range_obj.start.line
        start_char = range_obj.start.character
        end_line = range_obj.end.line
        end_char = range_obj.end.character
    else:
        start_line = start_char = end_line = end_char = 0
    
    story_symbol = create_document_symbol(
        name=story_name,
        construct_type="story_cycle",
        start_line=start_line,
        start_char=start_char,
        end_line=end_line,
        end_char=end_char,
        detail="Story Cycle",
    )
    
    # Walk children to find lifecycle hooks and effect groups
    if hasattr(story_cycle_node, 'children'):
        effect_group_count = 0
        for child in story_cycle_node.children:
            child_key = getattr(child, 'key', '')
            child_range = getattr(child, 'range', None)
            
            if not child_range:
                continue
            
            # Add lifecycle hooks as children
            if child_key in ('on_setup', 'on_end', 'on_owner_death'):
                hook_symbol = create_document_symbol(
                    name=child_key,
                    construct_type="lifecycle_hook",
                    start_line=child_range.start.line,
                    start_char=child_range.start.character,
                    end_line=child_range.end.line,
                    end_char=child_range.end.character,
                    detail="Lifecycle Hook",
                )
                story_symbol.children.append(hook_symbol)
            
            # Add effect groups as children
            elif child_key == 'effect_group':
                effect_group_count += 1
                effect_group_symbol = create_document_symbol(
                    name=f"effect_group_{effect_group_count}",
                    construct_type="effect_group",
                    start_line=child_range.start.line,
                    start_char=child_range.start.character,
                    end_line=child_range.end.line,
                    end_char=child_range.end.character,
                    detail="Effect Group",
                )
                story_symbol.children.append(effect_group_symbol)
    
    return story_symbol


def extract_document_symbols(parsed_document: Dict, file_path: str = None, schema_loader=None) -> List[DocumentSymbol]:
    """
    Extract all symbols from a parsed document.
    
    Now supports schema-driven extraction: if a schema is available for the file type,
    symbols are extracted according to the schema configuration. Otherwise, falls back
    to the original hardcoded extraction logic.

    Args:
        parsed_document: Parsed CK3 document (Dict or AST list)
        file_path: Optional file path for schema lookup
        schema_loader: Optional SchemaLoader for schema-driven extraction

    Returns:
        List of top-level DocumentSymbol objects
    """
    # Try schema-driven extraction first
    if file_path and schema_loader:
        try:
            from .schema_symbols import get_schema_symbols
            
            # If parsed_document is a list (AST), use it directly
            if isinstance(parsed_document, list):
                schema_symbols = get_schema_symbols(file_path, parsed_document, schema_loader)
                if schema_symbols:
                    # Convert LSP symbols back to internal format for compatibility
                    return [_lsp_to_internal(s) for s in schema_symbols]
        except Exception as e:
            # Fall back to hardcoded extraction on any error
            import logging
            logging.getLogger(__name__).debug(f"Schema symbol extraction failed: {e}")
    
    # Original hardcoded extraction logic (fallback)
    symbols = []

    # Extract events
    if isinstance(parsed_document, dict):
        if "events" in parsed_document:
            for event_node in parsed_document["events"]:
                symbols.append(extract_event_symbols(event_node))

        # Extract scripted effects
        if "scripted_effects" in parsed_document:
            for effect_node in parsed_document["scripted_effects"]:
                symbols.append(extract_scripted_effect_symbols(effect_node))

        # Extract scripted triggers
        if "scripted_triggers" in parsed_document:
            for trigger_node in parsed_document["scripted_triggers"]:
                symbols.append(extract_scripted_trigger_symbols(trigger_node))

        # Extract script values
        if "script_values" in parsed_document:
            for value_node in parsed_document["script_values"]:
                symbols.append(extract_script_value_symbols(value_node))

        # Extract on_actions
        if "on_actions" in parsed_document:
            for on_action_node in parsed_document["on_actions"]:
                symbols.append(extract_on_action_symbols(on_action_node))

    return symbols


def _lsp_to_internal(lsp_symbol: types.DocumentSymbol) -> DocumentSymbol:
    """Convert LSP DocumentSymbol back to internal format."""
    internal = DocumentSymbol(
        name=lsp_symbol.name,
        kind=lsp_symbol.kind,
        range=lsp_symbol.range,
        selection_range=lsp_symbol.selection_range,
        detail=lsp_symbol.detail
    )
    if lsp_symbol.children:
        internal.children = [_lsp_to_internal(child) for child in lsp_symbol.children]
    return internal


def convert_to_lsp_document_symbol(symbol: DocumentSymbol) -> types.DocumentSymbol:
    """
    Convert internal DocumentSymbol to LSP DocumentSymbol.

    Args:
        symbol: Internal DocumentSymbol object

    Returns:
        LSP DocumentSymbol object
    """
    children = [convert_to_lsp_document_symbol(child) for child in symbol.children]

    return types.DocumentSymbol(
        name=symbol.name,
        kind=symbol.kind,
        range=symbol.range,
        selection_range=symbol.selection_range,
        detail=symbol.detail,
        children=children if children else None,
    )


def find_symbols_by_name(symbols: List[DocumentSymbol], query: str) -> List[DocumentSymbol]:
    """
    Find symbols matching a query string.

    Used for workspace symbol search.

    Args:
        symbols: List of symbols to search
        query: Search query (case-insensitive)

    Returns:
        List of matching symbols
    """
    results = []
    query_lower = query.lower()

    for symbol in symbols:
        if query_lower in symbol.name.lower():
            results.append(symbol)

        # Recursively search children
        if symbol.children:
            results.extend(find_symbols_by_name(symbol.children, query))

    return results


def get_symbol_hierarchy(symbol: DocumentSymbol) -> str:
    """
    Get a hierarchical representation of a symbol.

    Used for displaying symbol context.

    Args:
        symbol: The symbol to represent

    Returns:
        String representation (e.g., "event_name > trigger")
    """
    parts = [symbol.name]

    # Could be extended to include parent information
    # if we track parent references

    return " > ".join(parts)

File: vscode-extension/src/server/schema/symbols.ts
/**
 * Schema-based Symbols - File-type-aware symbol extraction
 * 
 * This module extracts symbols from CK3 files based on schema definitions,
 * making symbol extraction file-type-aware and automatically supporting
 * new file types as schemas are added.
 * 
 * Responsibilities:
 * - Load symbol definitions from schemas for the current file type
 * - Extract symbols from AST based on schema patterns
 * - Provide hierarchical symbol structure
 * - Support nested symbols and symbol kinds
 */

import { DocumentSymbol, SymbolKind, Range } from 'vscode-languageserver';
import { CK3Node } from '../core/parser';
import { SchemaLoader, SchemaDefinition } from './loader';

/**
 * Schema Symbol Extractor extracts symbols based on schema definitions
 */
export class SchemaSymbolExtractor {
    private loader: SchemaLoader;

    constructor(loader: SchemaLoader) {
        this.loader = loader;
    }

    /**
     * Extract symbols from AST based on schema
     */
    public async extractSymbols(filePath: string, ast: CK3Node[]): Promise<DocumentSymbol[]> {
        const schema = await this.getSchemaForFile(filePath);
        if (!schema) {
            return [];
        }

        const symbols: DocumentSymbol[] = [];

        // Get symbol extraction configuration
        const symbolConfig = schema.symbols || {};
        const blockPattern = schema.identification?.block_pattern;

        for (const node of ast) {
            if (this.matchesBlockPattern(node.key, blockPattern)) {
                const symbol = this.createSymbol(node, schema, symbolConfig);
                if (symbol) {
                    symbols.push(symbol);
                }
            }
        }

        return symbols;
    }

    /**
     * Create a symbol from a node
     */
    private createSymbol(
        node: CK3Node,
        schema: SchemaDefinition,
        symbolConfig: any
    ): DocumentSymbol | null {
        // Determine symbol kind based on schema config
        const kind = this.getSymbolKind(node, symbolConfig);

        // Create symbol
        const symbol: DocumentSymbol = {
            name: node.key,
            detail: this.getSymbolDetail(node, schema),
            kind,
            range: node.range,
            selectionRange: node.range,
            children: []
        };

        // Extract nested symbols
        if (symbolConfig.nested) {
            symbol.children = this.extractNestedSymbols(node, schema, symbolConfig.nested);
        }

        return symbol;
    }

    /**
     * Extract nested symbols from a node
     */
    private extractNestedSymbols(
        node: CK3Node,
        schema: SchemaDefinition,
        nestedConfig: any
    ): DocumentSymbol[] {
        const symbols: DocumentSymbol[] = [];

        for (const child of node.children) {
            // Check if this child should be a symbol
            if (nestedConfig.fields && nestedConfig.fields.includes(child.key)) {
                const symbol = this.createNestedSymbol(child, nestedConfig);
                if (symbol) {
                    symbols.push(symbol);
                }
            }

            // Check for pattern-based symbols
            if (nestedConfig.pattern) {
                try {
                    const regex = new RegExp(nestedConfig.pattern);
                    if (regex.test(child.key)) {
                        const symbol = this.createNestedSymbol(child, nestedConfig);
                        if (symbol) {
                            symbols.push(symbol);
                        }
                    }
                } catch (error) {
                    // Invalid regex
                }
            }
        }

        return symbols;
    }

    /**
     * Create a nested symbol
     */
    private createNestedSymbol(node: CK3Node, config: any): DocumentSymbol | null {
        const kind = this.getNestedSymbolKind(node, config);

        return {
            name: node.key,
            detail: node.value ? String(node.value) : '',
            kind,
            range: node.range,
            selectionRange: node.range,
            children: []
        };
    }

    /**
     * Get symbol kind based on node and configuration
     */
    private getSymbolKind(node: CK3Node, config: any): SymbolKind {
        if (config.kind) {
            return this.mapSymbolKind(config.kind);
        }

        // Default kinds based on file type
        const fileName = node.key.toLowerCase();
        if (fileName.includes('event')) {
            return SymbolKind.Event;
        } else if (fileName.includes('decision')) {
            return SymbolKind.Function;
        } else if (fileName.includes('interaction')) {
            return SymbolKind.Interface;
        } else if (fileName.includes('on_action')) {
            return SymbolKind.Event;
        }

        return SymbolKind.Class;
    }

    /**
     * Get nested symbol kind
     */
    private getNestedSymbolKind(node: CK3Node, config: any): SymbolKind {
        // Map common field names to symbol kinds
        const kindMap: Record<string, SymbolKind> = {
            'option': SymbolKind.Method,
            'trigger': SymbolKind.Property,
            'immediate': SymbolKind.Property,
            'after': SymbolKind.Property,
            'effect': SymbolKind.Property,
            'desc': SymbolKind.String,
            'title': SymbolKind.String,
            'name': SymbolKind.String,
        };

        return kindMap[node.key] || SymbolKind.Field;
    }

    /**
     * Map symbol kind string to SymbolKind enum
     */
    private mapSymbolKind(kindStr: string): SymbolKind {
        const kindMap: Record<string, SymbolKind> = {
            'file': SymbolKind.File,
            'module': SymbolKind.Module,
            'namespace': SymbolKind.Namespace,
            'package': SymbolKind.Package,
            'class': SymbolKind.Class,
            'method': SymbolKind.Method,
            'property': SymbolKind.Property,
            'field': SymbolKind.Field,
            'constructor': SymbolKind.Constructor,
            'enum': SymbolKind.Enum,
            'interface': SymbolKind.Interface,
            'function': SymbolKind.Function,
            'variable': SymbolKind.Variable,
            'constant': SymbolKind.Constant,
            'string': SymbolKind.String,
            'number': SymbolKind.Number,
            'boolean': SymbolKind.Boolean,
            'array': SymbolKind.Array,
            'object': SymbolKind.Object,
            'key': SymbolKind.Key,
            'null': SymbolKind.Null,
            'event': SymbolKind.Event,
        };

        return kindMap[kindStr.toLowerCase()] || SymbolKind.Class;
    }

    /**
     * Get symbol detail text
     */
    private getSymbolDetail(node: CK3Node, schema: SchemaDefinition): string {
        const details: string[] = [];

        // Add file type
        if (schema.file_type) {
            details.push(schema.file_type);
        }

        // Add value if present
        if (node.value) {
            details.push(String(node.value));
        }

        return details.join(' - ');
    }

    /**
     * Check if a key matches the block identification pattern
     */
    private matchesBlockPattern(key: string, pattern?: string): boolean {
        if (!pattern) {
            return true;
        }
        try {
            const regex = new RegExp(pattern);
            return regex.test(key);
        } catch (error) {
            return false;
        }
    }

    /**
     * Get schema for file path
     */
    private async getSchemaForFile(filePath: string): Promise<SchemaDefinition | null> {
        const fileName = filePath.toLowerCase();

        if (fileName.includes('/events/') || fileName.includes('\\events\\')) {
            return await this.loader.loadSchema('events');
        } else if (fileName.includes('/decisions/') || fileName.includes('\\decisions\\')) {
            return await this.loader.loadSchema('decisions');
        } else if (fileName.includes('/character_interactions/') || fileName.includes('\\character_interactions\\')) {
            return await this.loader.loadSchema('character_interactions');
        } else if (fileName.includes('/on_actions/') || fileName.includes('\\on_actions\\')) {
            return await this.loader.loadSchema('on_actions');
        } else if (fileName.includes('/story_cycles/') || fileName.includes('\\story_cycles\\')) {
            return await this.loader.loadSchema('story_cycles');
        } else if (fileName.includes('/schemes/') || fileName.includes('\\schemes\\')) {
            return await this.loader.loadSchema('schemes');
        }

        return null;
    }
}
